### PR TITLE
security: validate hash of fetched binary

### PR DESCRIPTION
**Pull request by jesec**
Recently there are several high-profile supply chain attacks \[1\]\[2\]. This project, starred by over 15k developers with over 5k dependents, could be an ideal target of such attack.

Additionally, we ship prebuilt patched Node binaries, and users usually use them instead of compiling their own. This makes the problem worse, as it is more difficult to spot malicious binary. 

We are making progress by building binaries with a public CI service (Github Actions). The build processes are now transparent and auditable. Plus, even repo admins / org owners can not tamper the artifacts and logs of a Github Actions run. 

However, we maintainers do have the privileges to "Edit Release", and unfortunately, at the moment Github does not allow viewing the change history of release assets or locking down release assets. As a result, if one of us got compromised, the hacker can potentially seed malicious binaries to our users and end-users. 

Note that this is not a problem that is unique to us. Other projects such as [nexe](https://github.com/nexe/nexe/releases/latest) host binaries on Github, as well, and they have not deployed any defense. 

Currently, I post the GPG-signed hashes to the release page, so our users can validate binaries, and we can easily spot a compromise. However, most users probably wouldn't do that. And, this makes me the "trust anchor". While I take every precaution, I am not confident enough to say that I can defend my private key against state-sponsored hackers.

This change hardcodes the expected hashes of binaries and makes `pkg-fetch` validate hash of fetched binary. This mechanism makes the changes of binaries transparent and defeats compromise of control of release assets. 

With this change, the release flow:

1. Run binary workflows
2. Downloads binaries from Github Actions, gets hashes and commits the hashes
3. Bump the version, tag the revision, download binaries, upload binaries to Github Releases, release to npm

\[1\]: https://about.codecov.io/security-update
\[2\]: https://www.phoronix.com/scan.php?page=news_item&px=PHP-Git-Compromised

### CODE REVIEW

**Comment by leerob** [comment]
I love this write-up. Great explanation and brilliant idea.

File: lib/expected.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    'f06855896bfa10bead1f08fac080305fb6fbfb2cc691168a3f0f0e834e12bfba',
+  'node-v8.17.0-linux-x64':
+    '14d75d43de1ff86469d354bf42a83b9494e09502fa7bc23a975e2cb82b1608b0',
+  'node-v8.17.0-linuxstatic-arm64':
+    '84de8fe30b2bd1dcb3615cf1d1b538aa48e1fcf66620ef97dce6b7ae85b45025',
+  'node-v8.17.0-linuxstatic-x64':
+    '5206878079f160e75a02ad33b7559b4a869e8181ee03d51d7211b52995f9ca7b',
+  'node-v8.17.0-macos-x64':
+    'dffa71e39100f4daa57de73fda7b4debecd09f552b15cf11854c8475380d3817',
+  'node-v8.17.0-win-x64':
```
Comment (leerob) [inline lib/expected.ts:72]: Should we auto-generate these so it doesn't require a PR for every release? Or not doable?

Comment (jesec) [reply lib/expected.ts:72]: The whole point is to hardcode the hashes in the source code. So we do have to make a commit. 

You can get the hashes simply by `cat *.sha256sum` after downloading from Github Actions. Then, it is simple RegEx (to convert to object). 

To ease the process, I can make the CI print out the hashes in the log.

**Review by robertsLando — APPROVED** [review]
Well done @jesec ! LGTM

**Pull request merged**